### PR TITLE
Make upstart script more fault tolerant and easier to debug

### PR DESCRIPTION
**Pull request by ponychicken**
Fixes a number of smaller issues:
* Ensures that /etc/caddy exists
* Ensures correct HOME folder
* Adds logging capabilities

### CODE REVIEW

**Comment by wmark** [comment]
The ```cd /etc/caddy``` will change the working directory, yes, but other than that it has no effect: *Caddy* does not write anything to the *current directory* (whatever that might be). Certificates will still go to ```$HOME/.caddy```, so you either need to set that environment variable (→ `env`), or a global one for user ```www-data```.

Then there is the issue of *file descriptors*. On a webserver you don't want to run out of them. You need them whenever a file (for example, HTML or an image) is about to being opened. And, every open or lingering connection needs another. So better remove any limitations on that. (You did touch the file, now you own it including that line.)

Set Caddy's `-root=` to something safe (a directory that is guaranteed to be empty), or else—if the operator forgot to set one in his/her `Caddyfile`—you risk uncontrolled data exposure.

IMHO, the email address does not belong in the *init script* but in a `Caddyfile`.

File: dist/init/linux-upstart/caddy.conf [hunk 1]
```diff
@@ -14,5 +15,7 @@ respawn limit 10 5
 limit nofile 4096 4096
```
Comment (mholt) [inline dist/init/linux-upstart/caddy.conf]: This is equivalent to `ulimit -n` right? We should raise this as high as it can go, FWIW.

**Comment by mholt** [comment]
(Just my $0.02: How the email is specified is just a matter of preference. Putting it in the command as a flag makes it possible to distribute and use Caddyfiles without any modifications, for example.)

**Comment by wmark** [comment]
Init files could be shared among several installations or even as part of a distribution. They can be written in a generalized way.

If a Caddyfile need not list a domain name, then, I guess, it's when no *init file* will be used anyway.

**Comment by ponychicken** [comment]
Adding a dummy address ensures that the service will work without any additional config on the users side. I fixed the file limit and the env issue. Is /var/tmp safe as the root folder?

**Comment by mholt** [comment]
You can leave the email blank, don't use a dummy address. ;)

File: dist/init/linux-upstart/caddy.conf [hunk 2]
```diff
@@ -1,18 +1,25 @@
...
+env HOME=/etc/ssl/caddy
+
+limit nofile 1048576 1048576
 
 script
-	exec /usr/bin/caddy -agree=true -conf=/etc/caddy/Caddyfile
-end script
\ No newline at end of file
+	mkdir -p /etc/caddy
+	mkdir -p /var/tmp/caddy
```
Comment (wmark) [inline dist/init/linux-upstart/caddy.conf]: If you didn't find a safe directory (is there anything in `/proc/self` we could use?) then use a random directory like this:

```bash
mktemp -d -t "caddy-run.XXXXXX"
``` 

Please note that `/var/tmp` has stronger persistency guarantees than `/tmp`, hence it is an even less ideal location!

File: dist/init/linux-upstart/caddy.conf [hunk 3]
```diff
@@ -1,18 +1,25 @@
...
+	cd /etc/caddy
+	exec /usr/bin/caddy -agree=true -email -log=stdout -conf=/etc/caddy/Caddyfile -root=/var/tmp/caddy
```
Comment (wmark) [inline dist/init/linux-upstart/caddy.conf]: This is not how informed consent is supposed to be given. (I didn't even check if that works with an empty `-email`.)

Comment (mholt) [inline dist/init/linux-upstart/caddy.conf]: Incidentally (I didn't notice this before), but we can change `-agree=true` to just `-agree`. And the method of a command line flag is fine, I checked with them. Their own (former-official client) had the same thing.

**Comment by ponychicken** [comment]
I believed I addressed all the issues.

File: dist/init/linux-upstart/caddy.conf (hunk 3 above)
Comment (wmark) [inline dist/init/linux-upstart/caddy.conf]: I've been thinking about this scenario: »I didn't agree to anything. Oh, this is not me, it's been shipped this way by whoever maintains the PPA.«

IMHO the process needs the operator's consideration, and he needs to set at least one thing himself to show he understood that he is consenting into a binding agreement.

For Linux distros, *init* files should be close to immutable (some distros even ship them read-only! (like CoreOS)), and everything individual is expected to become part of configuration files.

File: dist/init/linux-upstart/caddy.conf [hunk 4]
```diff
@@ -1,18 +1,25 @@
...
+env HOME=/etc/ssl/caddy
+
+limit nofile 1048576 1048576
 
 script
-	exec /usr/bin/caddy -agree=true -conf=/etc/caddy/Caddyfile
-end script
\ No newline at end of file
+        mkdir -p /etc/caddy
+        rootdir=`mktemp -d -t "caddy-run.XXXXXX"`
```
Comment (wmark) [inline dist/init/linux-upstart/caddy.conf]: If it's a *bash* script (I don't know that) then use `$()`. The value should be quoted.

```bash
rootdir="$(mktemp -d -t "…")"
```

… Is there a way to run something when the process exits? If so, remove that directory there.

One more reason to like *systemd's* `PrivateTmp`. :-)

File: dist/init/linux-upstart/caddy.conf [hunk 5]
```diff
@@ -1,18 +1,25 @@
...
 
 setuid www-data
 setgid www-data
 
 respawn
 respawn limit 10 5
 
-limit nofile 4096 4096
+# Letsencrypt-issued certificates will be written to this directory.
+env HOME=/etc/ssl/caddy
```
Comment (mholt) [inline dist/init/linux-upstart/caddy.conf]: Just FYI, in the future, the .caddy folder may store more than just certificates, so it won't always be SSL-related. Also, Caddy will make a .caddy folder inside $HOME, so in this case it'd be `/etc/ssl/caddy/.caddy` - I'd just use an actual user HOME directory personally.

**Comment by mholt** [comment]
Thanks for the work @ponychicken! Almost there. :smile:

File: dist/init/linux-upstart/caddy.conf (hunk 5 above)
Comment (mholt) [inline dist/init/linux-upstart/caddy.conf]: @ponychicken Does this change sound reasonable to you? Sorry I wasn't clear.

File: dist/init/linux-upstart/caddy.conf [hunk 6]
```diff
@@ -1,18 +1,25 @@
...
+
+limit nofile 1048576 1048576
 
 script
-	exec /usr/bin/caddy -agree=true -conf=/etc/caddy/Caddyfile
-end script
\ No newline at end of file
+        mkdir -p /etc/caddy
```
Comment (weingart) [inline dist/init/linux-upstart/caddy.conf]: In general, at least in the "old days", startup scripts and such should not  modify files in `/etc/...`.  In some places, these filesystems are read-only.  Better options were:

- `/var/run/...`
- `/tmp`
- `/var/tmp`

Each with their own semantic differences (such as `/tmp` being wiped out at times, etc).

File: dist/init/linux-upstart/caddy.conf [hunk 7]
```diff
@@ -1,18 +1,25 @@
...
+        rootdir=`mktemp -d -t "caddy-run.XXXXXX"`
+        cd /etc/caddy
```
Comment (weingart) [inline dist/init/linux-upstart/caddy.conf]: I'd be tempted to either:

- `cd /`
- `cd /var/empty`

Depending on what is to be done.  More likely I'd `chdir("/")` in a daemon.

**Comment by mholt** [comment]
@ponychicken Have you had a chance to consider the last few suggestions/comments that have filtered in? Let's wrap this up.

File: dist/init/linux-upstart/caddy.conf (hunk 7 above)
Comment (weingart) [inline dist/init/linux-upstart/caddy.conf]: I'd personally go for `cd /`.  One of the main reasons is, if someone decides to `chroot(2)` the whole thing into whatever their env setup defaults are, they won't have to make unecessary changes or directories just to have things work...

**Comment by ponychicken** [comment]
@mholt I pushed a commit which fixes the quoting mentioned by @wmark .
The folder setup uses only /etc/caddy now, which existence is required already by looking for a caddy file there. I cd into it otherwise imports are broken (i'll open a bug for this)

**Comment by mholt** [comment]
Awesome - thanks for your contribution!! Hope to see you around more :smile:

**Comment by wmark** [comment]
LGTM

**Pull request merged**